### ISSUE TITLE
Vehicle acceleration seems lower and they turn very slowly with a jerky motion

### ISSUE THREAD

**Issue by paramat** [minetest/minetest#6907]
Try boats in MTG or my mesecar mod https://github.com/paramat/mesecar

**Comment by paramat** [comment]
Trying f4fedfed070ffd85c3446bc5d38d2fbd577640d6 before suspect 089f594582862d041a5bf3fa0f41d637248e31de

**Comment by paramat** [comment]
Suspect 089f594582862d041a5bf3fa0f41d637248e31de is OK.

**Comment by rubenwardy** [comment]
I suggest not interpolating entities players are attached to by default, as they're likely to be controlled by the player. Maybe this should be an entity property?

**Comment by paramat** [comment]
Trying 549cfd9db80c858bdc8d23a237ea57ccf5f68400

**Comment by paramat** [comment]
Trying e049405fdc688a4d619f3ae88008bbe5ee85b569

**Comment by paramat** [comment]
Trying 9c669016d1578a5c62f932c6ccb7a2b4b1e21f0a

**Comment by lhofhansl** [comment]
Do you see any warning like "active objects took ..."?

The normal server step is 0.09s, the active_object interval is 0.1, so that it won't fire each and every server step. But that should not cause a significant slowdown.

Hmm... Cannot have full speed and time budget. I introduced the budget for scenarios when active_object suck up a lot of CPU.

Can you try setting active_object_interval to 0.05 or something, and see what happens?
(You'll probably see a lot of warning like the one above, but see if that changes anything).

**Comment by lhofhansl** [comment]
Perhaps I should simple exclude player AOs from the budget and process them at full speed. Lemme try to a PR for this.

**Comment by paramat** [comment]
No warnings seen with any interval.
With `active_object_interval = 0.05` the motion is slightly improved but still slow and jerky.
> The normal server step is 0.09s, the active_object interval is 0.1, so that it won't fire each and every server step

That seems a bad idea, many objects rely on their step functions happening on every server step, this is what defines the speed of object behaviour.

**Comment by lhofhansl** [comment]
> many objects rely on their step functions happening on every server step

That's not true actually. The time difference is passed to the step function of each Lua object, that's what defines time. Otherwise we'd have different behavior on fast vs. slow machines.

The whole point of 9c66901 was to not step all AOs on every server step.

When active_object_interval is less than the server_step I don't see how this can possibly happen. Lemme check.

**Comment by lhofhansl** [comment]
Is `dedicated_server_step` not enforced?!

When I print dtime in ServerEnviroment::step I get about 1/60s - 1/30s.
Really? That'd be almost on each client frame.

Indeed... Looks like `Server::AsyncRunStep` is run at full throttle - slowed only by waiting to retrieve packets from the clients. So that explains the new jerkiness. Going from 0.016 to 0.1s is big change.

We can rollback the change... But... This change made things like NSSM playable for me. And most times calling LuaEntity.step that often is just wasted effort.

Now it seems that only really matters when the player is attached.
It's getting late, I'll look tomorrow.

**Comment by Calinou** [comment]
> When I print dtime in ServerEnviroment::step I get about 1/60s - 1/30s.
> Really? That'd be almost on each client frame.

Singleplayer uses a different value for server step, which usually runs as fast as possible.

**Comment by lhofhansl** [comment]
This was a local game. (not single player, but with a local server).
I found that AsyncRunStep is run at full speed and only slowed by receiving packets from the clients.

**Comment by lhofhansl** [comment]
OK... I tried various things, including taking both attached objects and players out of the loop and processing those at full speed. In all cases it leads to jerkiness.

Unless I come up with something soon, let's revert 9c66901

**Comment by lhofhansl** [comment]
I'm going to revert this. Speak up now if you disagree.

**Comment by lhofhansl** [comment]
This is will revert other things - like NSSM - to their old behavior.

If you try that, you'll see that at some point the world simply stops to be generated since all the time is spent in active objects. 9c66901 addressed that, and it allowed NSSM (and other heavy mob mods) to be playable under all conditions.

It seems we cannot easily have it both ways.

**Comment by lhofhansl** [comment]
@paramat can you try with active_object_interval set to 0.01 (which is < 1/60 s). That negates the aim of 9c66901, but it should restore the responsiveness.

If that works the easiest change is to default this to 0.01 (and get rid of the warning, which would be too spammy).

**Comment by stujones11** [comment]
I am not really sure you should be using NSSM as a fair representation of the average mod :)

I personally preferred the original behaviour in most cases, the fact that NSSM can have a server on it's knees is the fault of the mod IMO.

**Comment by lhofhansl** [comment]
You are not wrong :)

Perhaps it is best to revert this and then to regroup. I still want to find a way that allow for responsiveness in most cases, but degrades gracefully as the server gets busy.

**Comment by lhofhansl** [comment]
I'm ready to revert. As soon as I get an approval here from another member I'll push the revert.

**Comment by lhofhansl** [comment]
See [NSSM #22](https://github.com/NPXcoot/nssm/issues/22)

**Comment by paramat** [comment]
:+1: for revert, best option while we rethink.
